miner/miner_mev.go

```diff
@@ -0,0 +1,170 @@
+package miner
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/panjf2000/ants/v2"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	MevRoutineLimit              = 10000
+	TxDecodeConcurrencyForPerBid = 5
+	CheckTxDecodeInterval        = 3 * time.Millisecond
+)
+
+type BuilderConfig struct {
+	Address common.Address
+	URL     string
+}
+
+type MevConfig struct {
+	Enabled   bool            // Whether to enable Mev or not
+	SentryURL string          // The url of Mev sentry
+	Builders  []BuilderConfig // The list of builders
+
+	ValidatorCommission int64 // 100 means 1%
+}
+
+// MevRunning return true if mev is running.
+func (miner *Miner) MevRunning() bool {
+	return miner.bidSimulator.isRunning() && miner.bidSimulator.receivingBid()
+}
+
+// StartMev starts mev.
+func (miner *Miner) StartMev() {
+	miner.bidSimulator.startReceivingBid()
+}
+
+// StopMev stops mev.
+func (miner *Miner) StopMev() {
+	miner.bidSimulator.stopReceivingBid()
+}
+
+// AddBuilder adds a builder to the bid simulator.
+func (miner *Miner) AddBuilder(builder common.Address, url string) error {
+	return miner.bidSimulator.AddBuilder(builder, url)
+}
+
+// RemoveBuilder removes a builder from the bid simulator.
+func (miner *Miner) RemoveBuilder(builderAddr common.Address) error {
+	return miner.bidSimulator.RemoveBuilder(builderAddr)
+}
+
+func (miner *Miner) SendBid(ctx context.Context, bid *types.BidArgs) (common.Hash, error) {
+	builder, err := types.EcrecoverBuilder(bid)
+	if err != nil {
+		return common.Hash{}, types.NewInvalidBidError(fmt.Sprintf("invalid signature:%v", err))
+	}
+
+	bidTxs := make([]*BidTx, len(bid.Bid.Txs))
```

Comment:
seems there is no need to define a new type `BidTx`, just try an  error channel.